app/src/main/java/com/amaze/filemanager/database/UtilsHandler.java

```diff
@@ -114,6 +118,55 @@ public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
             case 1:
                 db.execSQL(querySftp);
                 break;
+            case 2:
+                String backupTable = "temp_" + TABLE_HISTORY;
+                String backupHistory = queryHistory.replace(TABLE_HISTORY, backupTable) +
+                        "INSERT INTO " + backupTable + " SELECT * FROM " + TABLE_HISTORY + " group by path;" +
+                        "DROP TABLE " + TABLE_HISTORY + ";" +
+                        "ALTER TABLE " + backupTable + " RENAME TO " + TABLE_HISTORY + ";";
+                db.execSQL(backupHistory);
+
+                backupTable = "temp_" + TABLE_HIDDEN;
+                String backupHidden = queryHistory.replace(TABLE_HIDDEN, backupTable) +
+                        "INSERT INTO " + backupTable + " SELECT * FROM " + TABLE_HIDDEN + " group by path;" +
+                        "DROP TABLE " + TABLE_HIDDEN + ";" +
+                        "ALTER TABLE " + backupTable + " RENAME TO " + TABLE_HIDDEN + ";";
+                db.execSQL(backupHidden);
+
+                backupTable = "temp_" + TABLE_LIST;
+                String backupList = queryHistory.replace(TABLE_LIST, backupTable) +
+                        "INSERT INTO " + backupTable + " SELECT * FROM " + TABLE_LIST + " group by path;" +
```

Comment:
Isn't this missing the "UNIQUE"?

Reply:
@VishalNehra @EmmanuelMess @TranceLove 
I test `backupTable` sql. 
``` java
            case 2:
                String backupTable = "temp_" + TABLE_HISTORY;
                String backupHistory = queryHistory.replace(TABLE_HISTORY, backupTable) +
                        "INSERT INTO " + backupTable + " SELECT * FROM " + TABLE_HISTORY + " group by path;" +
                        "DROP TABLE " + TABLE_HISTORY + ";" +
                        "ALTER TABLE " + backupTable + " RENAME TO " + TABLE_HISTORY + ";";
                Log.d("TAG", "onUpgrade: " + backupHistory);
```
It print follow sql
``` shell
D/TAG: onUpgrade: CREATE TABLE IF NOT EXISTS temp_history (_id INTEGER PRIMARY KEY,path TEXT UNIQUE);INSERT INTO temp_history SELECT * FROM history group by path;DROP TABLE history;ALTER TABLE temp_history RENAME TO history;
```
next, I test output sql in sqlite terminal
``` shell
Enter ".help" for usage hints.
sqlite> .schema history
CREATE TABLE history (_id INTEGER PRIMARY KEY,path TEXT);
sqlite> .table          
android_metadata  grid              history           sftp            
bookmarks         hidden            list              smb             
sqlite> select * from history;                                                                                                                                                                             1|/sdcard  
2|/sdcard
3|/storage/0C1B-2604
4|/storage/emulated/0
5|/storage/emulated/0/Movies
6|/storage/emulated/0
7|/storage/emulated/0/Movies
8|/storage/emulated/0
9|/storage/emulated/0/Notifications
10|/storage/emulated/0
11|/storage/emulated/0/Alarms
12|/storage/emulated/0
13|/storage/emulated/0/Android
14|/storage/emulated/0
15|/storage/emulated/0
16|/storage/0C1B-2604
17|/storage/emulated/0
sqlite> 
sqlite> CREATE TABLE IF NOT EXISTS temp_history (_id INTEGER PRIMARY KEY,path TEXT UNIQUE);INSERT INTO temp_history SELECT * FROM history group by path;DROP TABLE history;ALTER TABLE temp_history RENAME TO history;
sqlite> 
sqlite> .table
android_metadata  grid              history           sftp            
bookmarks         hidden            list              smb             
sqlite> .schema history
CREATE TABLE IF NOT EXISTS "history" (_id INTEGER PRIMARY KEY,path TEXT UNIQUE);
sqlite> select * from history;
2|/sdcard
7|/storage/emulated/0/Movies
9|/storage/emulated/0/Notifications
11|/storage/emulated/0/Alarms
13|/storage/emulated/0/Android
16|/storage/0C1B-2604
17|/storage/emulated/0
sqlite> 
```
When I upgrade database. can't upgrade. I am so confuse
``` shell
sqlite> .schema
CREATE TABLE android_metadata (locale TEXT);
CREATE TABLE history (_id INTEGER PRIMARY KEY,path TEXT);
CREATE TABLE hidden (_id INTEGER PRIMARY KEY,path TEXT);
CREATE TABLE list (_id INTEGER PRIMARY KEY,path TEXT);
CREATE TABLE grid (_id INTEGER PRIMARY KEY,path TEXT);
CREATE TABLE bookmarks (_id INTEGER PRIMARY KEY,name TEXT,path TEXT);
CREATE TABLE smb (_id INTEGER PRIMARY KEY,name TEXT,path TEXT);
CREATE TABLE sftp (_id INTEGER PRIMARY KEY,name TEXT,path TEXT UNIQUE,pub_key TEXT,ssh_key_name TEXT,ssh_key TEXT);
CREATE TABLE temp_history (_id INTEGER PRIMARY KEY,path TEXT UNIQUE);
sqlite> select * from history;
1|/sdcard
2|/sdcard
3|/storage/0C1B-2604
4|/storage/emulated/0
5|/storage/emulated/0/Alarms
6|/storage/emulated/0
7|/storage/emulated/0/Alarms
8|/storage/emulated/0
9|/storage/emulated/0/Android
10|/storage/emulated/0
11|/storage/emulated/0/Android
12|/storage/emulated/0
13|/storage/emulated/0/DCIM
14|/storage/emulated/0
15|/storage/emulated/0/DCIM
16|/storage/emulated/0
17|/storage/emulated/0
18|/storage/0C1B-2604
sqlite>               
sqlite> select * from temp_history;
sqlite> 
```
Help me find out what problem i made. thanks in advance

Reply:
Why do you group by path in `INSERT INTO temp_history SELECT * FROM history group by path;`?

Reply:
`group by` can remove duplicate row

Reply:
Do you mean should be like this ?
``` java
case 1:
    db.execSQL(querySftp);
     //break;
case 3:
      ....
     break;
```

Reply:
No,

```
case 1:
    db.execSQL(querySftp);
     //break;
case 2:
      ....
     //break;
```

Reply:
change this will work ?

Reply:
What do you mean?

Reply:
Logically, when `oldVersion` is 1, both `case 1:` and `case 2:` will run because of no `break`s.
When `oldVersion` is 2, only `case 2:` will run.

So code-wise, yes.

Reply:
@TranceLove @VishalNehra @EmmanuelMess I update code. please review again :)